### PR TITLE
Support ACR in Anonymized Export

### PR DESCRIPTION
**Pull request by yankunhuang-pku**
## Description
In this PR, we add a new feature of anonymized export: 
- Customer can export anonymized data with customized anonymization configuration on Azure Container Registry (ACR).

## Related issues
Work item [#90039](https://microsofthealth.visualstudio.com/Health/_workitems/edit/90039).

## Testing
- Unit Tests for different components such as Anonymizer and ExportAnonymizerFactory.
- E2E tests at API level for R4 and STU3. Also add E2E test cases for negative cases.

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [x] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [x] CI is green before merge
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Api.UnitTests/Controllers/ExportControllerTests.cs [hunk 1]
```diff
@@ -112,38 +127,102 @@ public async Task GivenAnExportResourceTypeIdRequest_WhenResourceTypeIsNotGroup_
...
+                resourceType: null,
+                containerName: _testContainer,
+                formatName: null,
+                anonymizationConfigCollectionReference: _testAnonymizationConfigCollectionReference,
+                anonymizationConfigLocation: _testConfig,
+                anonymizationConfigFileETag: _testAnonymizationConfigEtag,
+                typeParameter: ResourceType.Patient.ToString()));
+        }
+
+        [Theory]
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Api.UnitTests/Controllers/ExportControllerTests.cs:195]: Could you add a comment with the valid collection reference so that it is more clear what error each test case is testing?

File: src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+using Microsoft.Health.Fhir.TemplateManagement.ArtifactProviders;
+using Microsoft.Health.Fhir.TemplateManagement.Client;
+using Microsoft.Health.Fhir.TemplateManagement.Exceptions;
+using Microsoft.Health.Fhir.TemplateManagement.Models;
+using Microsoft.Health.Fhir.TemplateManagement.Utilities;
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Fhir.Azure
+{
+    public class AnonymizationConfigurationArtifactProvider : IArtifactProvider, IDisposable
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs:34]: Would it be possible to add unit tests for this class? I know the old Export Destination Artifact Provider didn't have any, but I think they would be useful.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+
+            foreach (string content in dataFromExport)
+            {
+                Resource result = parser.Parse<Resource>(content);
+
+                Assert.Contains(result.Meta.Security, c => "REDACTED".Equals(c.Code));
+            }
+
+            // Only check metric for local tests
+            if (_isUsingInProcTestServer)
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs:105]: This if will never be true because the test would have been skipped on line 79.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+            string reference = $"{registry.Server}/{TestRepositoryName}:{TestRepositoryTag}";
+            Uri contentLocation = await _testFhirClient.AnonymizedExportUsingAcrAsync(TestConfigName, reference, dateTime, containerName);
+            HttpResponseMessage response = await WaitForCompleteAsync(contentLocation);
+            string responseContent = await response.Content.ReadAsStringAsync();
+
+            Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+            Assert.Contains("Failed to parse configuration file", responseContent);
+
+            // Only check metric for local tests
+            if (_isUsingInProcTestServer)
```
[the remark this answers is not in the document]
Comment (LTA-Thinking) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs:225]: There is one of these in every test that should be removed.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+
+                response = await _testFhirClient.CheckExportAsync(contentLocation);
+
+                resultCode = response.StatusCode;
+            }
+
+            return response;
+        }
+
+        private async Task<IList<Uri>> CheckExportStatus(HttpResponseMessage response)
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs:306]: For CheckExportStatus, DownloadBlobAndParse, and GetCloudStorageAccountHelper is there enough unique requirements for this test class to rewrite them or could the helper methods in ExportTestHelper be used?

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+using FhirGroup = Hl7.Fhir.Model.Group;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.Tests.E2E.Rest
+{
+    [Trait(Traits.Category, Categories.AnonymizedExport)]
+    [HttpIntegrationFixtureArgumentSets(DataStore.All, Format.Json)]
+    public class AnonymizedExportUsingAcrTests : IClassFixture<ExportTestFixture>
+    {
+        private const string TestRepositoryName = "anonymizationconfigs";
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs:46]: Can we get the repository name from an environment variable instead of hardcoding it into the test file? We shouldn't be linking our code to specific Azure resources.

File: src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+            _exportJobConfiguration = exportJobConfiguration.Value;
+            _logger = logger;
+
+            _cache = new MemoryCache(new MemoryCacheOptions
+            {
+                SizeLimit = _exportJobConfiguration.CacheSizeLimit,
+            });
+        }
+
+        public async Task FetchAsync(string location, Stream targetStream, CancellationToken cancellationToken)
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs:69]: Can the location string be changed to be an ExportJobRecord object? That would save us having to serialize and deserialize it several times.

File: src/Microsoft.Health.Fhir.Shared.Api.UnitTests/Controllers/ExportControllerTests.cs (hunk 1 above)
Comment (yankunhuang-pku) [reply src/Microsoft.Health.Fhir.Shared.Api.UnitTests/Controllers/ExportControllerTests.cs:199]: Sure~Added.

File: src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs (hunk 2 above)
Comment (yankunhuang-pku) [reply src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs:34]: OK, I've tried to add some unit tests for it.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yankunhuang-pku) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/AnonymizedExportUsingAcrTests.cs]: Thanks~ Removed.

File: src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs (hunk 7 above)
Comment (yankunhuang-pku) [reply src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs:69]: The purpose of this implementation is to be consistent with the defined interface `IArtifactProvider`. I think type `string` woule be more general.
```
public interface IArtifactProvider
{
    /// <summary>
    /// Fetch artifect used by FHIR server.
    /// </summary>
    /// <param name="location">The location of the artifect. The location string format depends on the artifact provider.</param>
    /// <param name="targetStream">The stream for target artifact content.</param>
    /// <param name="cancellationToken">The cancellation token.</param>
    Task FetchAsync(string location, Stream targetStream, CancellationToken cancellationToken);
}
```

File: src/Microsoft.Health.Fhir.Azure.UnitTests/AnonymizationConfigurationArtifactProviderTests.cs [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+                stream.Position = 0;
+                using (StreamReader reader = new StreamReader(stream))
+                {
+                    string configurationContent = await reader.ReadToEndAsync();
+                    Assert.Contains("Resource.nodesByName('id')", configurationContent);
+                }
+            }
+        }
+
+        [Fact]
```
[the remark this answers is not in the document]
Comment (yankunhuang-pku) [reply src/Microsoft.Health.Fhir.Azure.UnitTests/AnonymizationConfigurationArtifactProviderTests.cs]: Thanks~ Updated.

File: src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs (hunk 7 above)
Comment (yankunhuang-pku) [reply src/Microsoft.Health.Fhir.Azure/AnonymizationConfigurationArtifactProvider.cs]: Yes, make sense if the method won't work anymore unless the string is an ExportJobRecord. 
I've changed the interfaces.

**Review by LTA-Thinking — APPROVED** [review]
